csrc/fusion_segmenter.cpp

```diff
@@ -3715,6 +3715,125 @@ class MergeUpAndDownCast {
   SegmentCandidateFinder* segment_candidate_finder_ = nullptr;
 };
 
+// Concat is represented as PadOp nodes of inputs, followed by a
+// ConcatOp node. It's unlikely they should be separated.
+class MergeCatWithInputPads {
+ public:
+  static void run(SegmentCandidateFinder* segment_candidate_finder) {
+    MergeCatWithInputPads merge_cat_with_input_pads(segment_candidate_finder);
+  }
+
+ private:
+  MergeCatWithInputPads(SegmentCandidateFinder* segment_candidate_finder)
+      : segment_candidate_finder_(segment_candidate_finder) {
+    merge();
+  }
+
+  void merge() {
+    std::unordered_set<SegmentedGroup*> considered_groups;
+    std::vector<std::vector<SegmentedGroup*>> candidates;
+
+    for (SegmentedGroup* group : segment_candidate_finder_->groups()) {
+      if (considered_groups.contains(group)) {
+        continue;
+      }
+
+      auto groups_to_merge = getGroupsToMerge(group);
+      if (!groups_to_merge.has_value()) {
+        continue;
+      }
+
+      considered_groups.insert(
+          groups_to_merge->begin(), groups_to_merge->end());
+
+      candidates.emplace_back(*groups_to_merge);
+    }
+
+    // Try merging the detected group
+    for (const auto& groups_to_merge : candidates) {
+      mergeGroups(groups_to_merge);
+    }
+  }
+
+  // Given a segmented group, try to detect a concat pattern, where
+  // all of the inputs are produced by pad groups.
+  std::optional<std::vector<SegmentedGroup*>> getGroupsToMerge(
+      SegmentedGroup* cat_group) {
+    // This pass is assumed to be applied before the iterative
+    // merge step, so only consider groups that are not yet merged
+    // at all.
+    if (cat_group->exprs().size() != 1) {
+      return std::nullopt;
+    }
+
+    // Technically, this can be just an add operation as concat can be
+    // represented with pad and add.
+    // TODO: Consider extending this pattern matching to support
+    // add-based concat
+    auto cat = dynamic_cast<CatOp*>(cat_group->exprs().at(0));
+    if (cat == nullptr) {
+      return std::nullopt;
+    }
+
+    // Check if a given pad is for the cat. More strictly, the actual
+    // padding widths do matter, but it's unikely to make any
+    // difference. Since this is a heuristic, this check should be
+    // good enough.
+    auto is_matching_pad = [&cat](PadOp* pad) {
+      if (!pad->value()->isZero()) {
+        return false;
+      }
+      auto padded_axes = pad->getPaddedAxes();
+      return padded_axes.size() == 1 &&
+          padded_axes.at(0) == cat->concatenatedDim();
+    };
+
+    std::vector<SegmentedGroup*> groups_to_merge;
+    groups_to_merge.reserve(cat->inputs().size() + 1);
+    for (const auto cat_inp : cat->inputs()) {
+      auto pad = dynamic_cast<PadOp*>(cat_inp->definition());
+      if (pad == nullptr || !is_matching_pad(pad)) {
+        return std::nullopt;
```

Comment:
QQ: do we support pattern of mismatched pad? i.e. only a subset of inputs to cat is being padded.

I wonder if we should do `continue` instead.

Reply:
When would that happen?

Reply:
Looks like our c++ API already assert on that. https://github.com/NVIDIA/Fuser/blob/49b52700966c3c3e3da291e37058779e3b845c2e/csrc/ops/alias.cpp#L638-L645

I'm wondering if this segmentation logic would also be beneficial to add-based cat ops. We can probably manually construct something like that -> thinking about optimization passes that would reorder some of the pad across its producer...